spec/index.bs

```diff
@@ -1338,20 +1344,112 @@ Only supported capabilities may be requested in {{GPUAdapter/requestDevice()}};
+If any feature level is requested on an implementation that only supports lower feature levels,
+the implementation must return `null`.
+
+Implementations must not automatically upgrade feature level requests, unless the requested
+feature level is <dfn dfn for=GPUFeatureLevel>upgradeable</dfn> to another feature level.
+<!-- This is defined on a level-by-level basis. For example, someday, FL1 could allow upgrading to
+FL2, if we want to allow implementation to drop their implementations of FL1 validation entirely. -->
+
+<script type=idl>
+enum GPUFeatureLevel {
+    "compatibility",
+    "fl1",
+    "fl1-desktop",
+};
+</script>
+
+<dl dfn-type=enum-value dfn-for=GPUFeatureLevel>
+    : <dfn>"compatibility"</dfn>
+    ::
+        Issue: remove for initial landing
+
+        This is the minimum feature level.
+
+        It is [=GPUFeatureLevel/upgradeable=] to {{GPUFeatureLevel/"fl1"}}.
+
+        It provides the following device capabilities by default:
+
+        - Limits as described in the [=supported limits=] table.
+
+        It also guarantees of any adapter returned:
+
+        - At least one of the following:
+            - {{GPUFeatureName/"texture-compression-bc"}} is supported.
+            - {{GPUFeatureName/"texture-compression-etc2"}} is supported.
+
+    : <dfn>"fl1"</dfn>
+    ::
+        Issue: bikeshed name
+
+        This feature level is the default for
+        {{GPURequestAdapterOptions}}.{{GPURequestAdapterOptions/minFeatureLevel}}.
+
+        It provides the following device capabilities by default:
+
+        - Limits as described in the [=supported limits=] table.
+        - {{GPUFeatureName/"everything-from-compat-to-fl1"}}
+
+        It also guarantees of any adapter returned:
+
+        - At least one of the following:
+            - {{GPUFeatureName/"texture-compression-bc"}} is supported.
+            - Both {{GPUFeatureName/"texture-compression-etc2"}} and
+                {{GPUFeatureName/"texture-compression-astc"}} are supported.
+
+    : <dfn>"fl1-desktop"</dfn>
+    ::
+        Issue: Just an example, probably not something we want to do exactly. Remove for initial landing.
+
+        This feature level describes "desktop" class hardware.
+
+        It provides the following device capabilities by default:
+
+        - Limits as described in the [=supported limits=] table.
+        - {{GPUFeatureName/"texture-compression-bc"}}
+</dl>
+
 ### Features ### {#features}
 
 A <dfn dfn>feature</dfn> is a set of optional WebGPU functionality that is not supported
 on all implementations, typically due to hardware or system software constraints.
 
-Functionality that is part of a feature may only be used if the feature was requested at device
-creation (in {{GPUDeviceDescriptor/requiredFeatures}}).
+Functionality that is part of a feature may only be used if the feature was enabled at device
+creation; either:
+
+- Part of the adapter's {{adapter/[[default feature level]]}}
```

Comment:
Hm... maybe I don't understand what the `[[default feature level]]` is supposed to represent. I thought it was the feature level that everything would be upgraded to at minimum, but that doesn't make sense in this context.

So I guess the adapter has two feature level values: The internal minimum, which everything will be upgraded to as a baseline to reduce implementation tech debt, and the requested feature level passed to `requestAdapter()`, which will presumably be a few levels above the minimum?  I now guess that `[[default feature level]]` is the latter, but I don't think the name makes that particularly clear. And I think we'll need a slot for the minimum outside of the value passed in the dictionary.

This also makes me think of an interesting question: What benefit is there to supporting anything other than feature level {max} once as an implementor once our implementation catches up to that point? If I'm allowed to auto upgrade requests is there any benefit for me to keep advertising a minimum feature level of {max - 2}? Seems like there's less tech debt in it for me if I can just auto-upgrade everyone the moment the feature level has been defined?

Reply:
> What benefit is there to supporting anything other than feature level {max} once as an implementor once our implementation catches up to that point?

☝️ Hi, this is me, a GPU developer who someone temporarily forgot that _hardware differences exist_. Please ignore this question. 😓

Reply:
It is the feature level of the adapter request (after any auto-upgrading that the implementation has done because it doesn't support the originally-requested feature level).
This feature level defines the default set of capabilities that you'll get from any requestDevice() call on the adapter - even if you attempt to request worse capabilities.

The motivation is:
- It creates a path for implementations to auto-upgrade in a backward-compatible way (e.g. from compat to fl1)
- Nothing after `requestAdapter()` needs to care about the `minFeatureLevel`.
- Nothing after `requestDevice()` needs to care about the `[[default feature level]]`.